bcgov/GDX-Analytics-microservice
# Gdxdsd 7197 cmslite user data to redshift split

### This PR does the following:

1. Moves and/or copies the python scripts cmslite_user_data_to_redshift, along with its associated config, ddl, and development files into a separate cmslite_user_data__to_redshift folder
2. Creates new python virtual environment files that only use packages specific to cmslite_user_data_to_redshift
3. Edits the README.md file to removed details specific to the s3_to_redshift, asset_data_to_redshift, and build_derived_assets microservices, and add details about cmslite_user_data_to_redshift

Please note that some files have been modified for testing purposes. You can view the modified files in the ticket

### Login to the development environment:

1. Run the following command to login to EC2
```
$ awsmfa prod ######
Generating new IAM STS Token ...
STS Session Token generated and saved in profile mfa successfully.
$ microservice_ssm
```
2. Change the directory to the development branch:
```
$ cd /home/microservice/branch/GDXDSD-7197_cmslite_user_data_to_redshift_split/cmslite_user_data_to_redshift
```

### Review README.md
Review [README.md](https://github.com/bcgov/GDX-Analytics-microservice/blob/GDXDSD-7197_cmslite_user_data_to_redshift_split/cmslite_user_data_to_redshift/README.md) and check to makes sure there are no details specific to s3_to_redshift, asset_data_to_redshift, and build_derived_assets. Also make sure the details about cmslite_user_data_to_redshift.py makes sense. 
Please note that there are possibly still references to these other microservices where relevant. But these references shouldn't be related to how the microservice works or how the configs are structured

### Review moved files

Part of this changes includes moving files from the old combined folder into a separate one specific for this microservice. In this section, we will be reviewing to make sure all of these files were moved successfully
1. Review [cmslite_user_data_to_redshift](https://github.com/bcgov/GDX-Analytics-microservice/tree/GDXDSD-7197_cmslite_user_data_to_redshift_split/cmslite_user_data_to_redshift) to make sure the microservice scripts have been moved successfully
- cmslite_user_data_to_redshift.py
3. Run the following command to get a list of the scheduled microservice's configs:
```
$ crontab -l | grep cmslite_user_data_to_redshift
```
4. Review the files in [config.d](https://github.com/bcgov/GDX-Analytics-microservice/tree/GDXDSD-7197_cmslite_user_data_to_redshift_split/cmslite_user_data_to_redshift/config.d) and make sure any config that appeared in the previous step is listed
5. Review the files in [ddl](https://github.com/bcgov/GDX-Analytics-microservice/tree/GDXDSD-7197_cmslite_user_data_to_redshift_split/cmslite_user_data_to_redshift/ddl) and make sure any table listed in one of the previous step's configs has an associated DDL

### Review copied files

1. Review the files in [cmslite_user_data_to_redshift](https://github.com/bcgov/GDX-Analytics-microservice/tree/GDXDSD-7197_cmslite_user_data_to_redshift_split/cmslite_user_data_to_redshift) to make sure the following file exist
- .flake8
- pylintrc
- README.md

### Review created files
1. Review the files in [cmslite_user_data_to_redshift](https://github.com/bcgov/GDX-Analytics-microservice/tree/GDXDSD-7197_cmslite_user_data_to_redshift_split/cmslite_user_data_to_redshift) to make sure the following file exist
- Pipfile
- Pipfile.lock

### Testing the microservice:
1. Run the following command to create the test tables:
```
$ microservice_rs < ddl/GDXDSD-7197_cmslite.group.sql 
SET
CREATE TABLE
GRANT
$ microservice_rs < ddl/GDXDSD-7197_cmslite.user_activity.sql 
SET
CREATE TABLE
GRANT
$ microservice_rs < ddl/GDXDSD-7197_cmslite.user_group.sql 
SET
CREATE TABLE
GRANT
$ microservice_rs < ddl/GDXDSD-7197_cmslite.user_status.sql 
SET
CREATE TABLE
GRANT
```
2. Run the following command to test the cmslite_user_data_to_redshift.py script and compare it to the expected output:
```
$ /home/microservice/.local/bin/pipenv run python3.8 cmslite_user_data_to_redshift.py config.d/GDXDSD-7197_cmslite_user_group_metadata.json
Report cmslite_user_data_to_redshift.py:

Config: config.d/GDXDSD-7197_cmslite_user_group_metadata.json

Microservice started at: 2024-11-14 15:56:42-0800 (PST), ended at: 2024-11-14 15:57:49-0800 (PST), elapsing: 0:01:06.652056.

Objects to process: 1
Objects successfully processed: 1
Objects that failed to process: 0
Objects output to 'processed/good': 1
Objects output to 'processed/bad': 0
Tables loaded to Redshift: 4/4

List of objects successfully fully ingested from S3, processed, loaded to S3 ('good'), and copied to Redshift:
client/doug_test/GDXDSD-7197/cmslite_gdx/cms-analytics-csv.2024-11-14.tgz

List of tables that were successfully loaded into Redshift:
test.gdxdsd7197_cms_group
test.gdxdsd7197_user_activity
test.gdxdsd7197_user_group
test.gdxdsd7197_user_status
```
3. Run the following query to check if the data was loaded correctly:
```
$ microservice_rs
SET
snowplow=> select count(*) from test.gdxdsd7197_user_group;
 count 
-------
  8168
(1 row)

snowplow=> select count(*) from test.gdxdsd7197_user_status;
 count 
-------
  5903
(1 row)

snowplow=> select count(*) from test.gdxdsd7197_user_activity;
 count 
-------
 66520
(1 row)

snowplow=> select count(*) from test.gdxdsd7197_cms_group;
 count 
-------
   817
(1 row)

snowplow=> \q
```
4. Check the following files appear in S3 batch: https://ca-central-1.console.aws.amazon.com/s3/buckets/sp-ca-bc-gov-131565110619-12-microservices?region=ca-central-1&bucketType=general&prefix=processed/batch/client/doug_test/GDXDSD-7197/cmslite_gdx/&showversions=false
- analytics_group.csv.2024-11-14
- analytics_user_activity.csv.2024-11-14
- analytics_user_group.csv.2024-11-14
- analytics_user_status.csv.2024-11-14
5. Check if the following file appears in S3 good: https://ca-central-1.console.aws.amazon.com/s3/buckets/sp-ca-bc-gov-131565110619-12-microservices?region=ca-central-1&bucketType=general&prefix=processed/good/client/doug_test/GDXDSD-7197/cmslite_gdx/&showversions=false
- cms-analytics-csv.2024-11-14.tgz
6. Delete the following files from S3 batch: https://ca-central-1.console.aws.amazon.com/s3/buckets/sp-ca-bc-gov-131565110619-12-microservices?region=ca-central-1&bucketType=general&prefix=processed/batch/client/doug_test/GDXDSD-7197/cmslite_gdx/&showversions=false
- analytics_group.csv.2024-11-14
- analytics_user_activity.csv.2024-11-14
- analytics_user_group.csv.2024-11-14
- analytics_user_status.csv.2024-11-14
7. Delete the following file from S3 good: https://ca-central-1.console.aws.amazon.com/s3/buckets/sp-ca-bc-gov-131565110619-12-microservices?region=ca-central-1&bucketType=general&prefix=processed/good/client/doug_test/GDXDSD-7197/cmslite_gdx/&showversions=false
- cms-analytics-csv.2024-11-14.tgz
8. Run the following queries to drop the test tables:
```
$ microservice_rs
SET
snowplow=> drop table test.gdxdsd7197_user_group;
DROP TABLE
snowplow=> drop table test.gdxdsd7197_user_status;
DROP TABLE
snowplow=> drop table test.gdxdsd7197_user_activity;
DROP TABLE
snowplow=> drop table test.gdxdsd7197_cms_group;
DROP TABLE
snowplow=> \q
```
